### ISSUE TITLE
server: enable SQL clients to log in with a token

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#65607]
This stems from a conversation with @jordanlewis last week.

We would like to enable a user journey that looks like this:

1. (once) user creates account using a web form
2. (once) user runs `cloud login` command in their terminal. This opens an authentication flow as is customary with other cloud technologies nowadays:
   - if the command can invoke a web browser, then it does so and points it to the URL that starts the flow.
   - if the command is run remotely, it prints out a URL and instructs the user to copy-paste it into a web browser
   - then, asynchronously, the command polls the server-side for completion of the login flow (or a timeout)
   - the login flow is a web app that optionally integrates with oAuth to log in the user, and when the login flow completes, stores a token in a place that the client-side command can retrieve.
   - when the client-side command finds the token server-side, it retrieves it and stores it client-side in a filesystem location that is going to be the same for all client apps (decided by convention, for example `~/.config/XXXX` somewhere)

3. every time a client app starts, it can then look up the token in that agreed-upon location in the filesystem, and use that to construct connection parameters to the service. There is further no configuration needed.

   Example login flows for client apps:

   - using `cockroach sql`:

     ```
     $ cockroach sql   # notice -- no extra arguments
     ...
     user@serveraddr:/defaultdb>  -- user can enter SQL here.
     ```

   - using a Python program:

     ```
     import cockroach
     connmgr = cockroach.open() # no args! see below for details
     conn = connmgr.connect() # no args
     # from here on the conn object is a regular SQL connection object
     ```

## What's happening behind the scenes

What happens behind the scene for the creation of a token is bog-standard with cloud technologies, details left as an exercise for the reader. Eventually, the token makes its way to the filesystem, for example in `~/.config/cockroachcloud/token`

Additionally, the login command can be responsible for retrieving the CA certificate to validate the server's identity during subsequent client conns, when the service does not use a standard, public CA or when client drivers don't know how to use standard CAs (this is sadly the case for many SQL client drivers).

What's in a token exactly?

- the token is a string that proves that the user's unix account has access to the service.
  the token also contains enough information to help the server decide which user is using the service, when the token is delivered on the other side.

  conceptually, a token must contain at least:
  - the identity of the user
  - their authentication credential

  (This is similar to [jwts (JSON web tokens)](https://en.wikipedia.org/wiki/JSON_Web_Token) - although we're unlikely to want to use jwts because the SQL protocol wants us to use a more compact token for client logins)

- additionally, each client command needs to know where to find the service on the network. How does it do that?
 
  - one way is to agree ahead of time that the service is _always_ going to be reachable via a fixed DNS name, for example `sql.cockroachcloud.com`. In that view, client apps _always_ establish their client connection via this domain name.

    pros: "simple" to implement client-side, programmers who want to do things "manually" can be instructed how to do this in few steps.
    cons: a single service (load balancer) is responsible for routing the requests from all clients, across all apps. Even if that service is replicated to handle more load, each instance of it needs access to a directory that maps tokens to specific instances of the service (behind the load balancer). This assigns a lot of responsibility for that one service. This makes it a single point of failure, and also a target for security attacks.

  - another way is to embed the address of the target service inside the token! This way, which network server to use is decided during the do-once login step, and that address is stored alongside the other parts of the token for reuse by client drivers.

    pros: more robust, and it gives the ability of the login system to hand out tokens for multiple services (for Cockroachcloud this is advantageous, because it makes it able to deliver tokens for dedicated clusters which 1) do not have a login proxy for them 2) where users have an expectation that they have "direct" access to their crdb nodes, not shared with other users)
    cons: the login system must deliver one extra bit of information in the login tokens.

So to summarize a login token could contain the following information:

- network address of the service's entry point (in CC dedicated that would be the dedicated cluster's LB, for shared tier the SQL proxy)
- identity of the user and target instance (SQL user + in CC shared tier that would be the CC user ID and their tenant ID)
- proof of access: an authentication token

Once this information is available in the filesystem, it is sufficient for a client driver to creates a client connection and access the service.

### What's in an authentication token

One naive form of authn token is a login password.
Pros: simple
Cons: not very secure
Cons: also this is expensive to validate server-side: every time the client app connects, the server must perform a lookup in the password database to determine whether the password is valid.

Another form of authn token is a TLS client certificate.
Pros: more secure.
Pros: the server does not need to perform a databse lookup; it can simply check that the CA for the client cert is the one it knows about already.

### How to think about revocation, and token renewal

An authn credential that's valid forever is not too good, because it becomes too easy to misplace it and grant access to malicious users.

However if we make it temporary, we run into another problem: a client app  that runs for a long time (e.g. a microservice that intends to run for years) can "exhaust" its token and lose access when the token expires.

What to do about this?

The client-side framework that manages connections must be responsible for "refreshing" the authentication token when it's about to expire.

How does this work? As long as a connection is established, in the background the framework must poll the server every now and then (e.g. every hour) to refresh the token and optionally retrieve a new token if the server wants to deliver a new one.

This way the server can also decide when to block access to a client who has misbehaved or perhaps whose credentials seem to have been stolen by a 3rd party.

This type of refresh mechanism is good for any kind of auhn token - passwords, API keys, TLS client certs etc.

## What components are needed?

- a login component: something that takes a URL from a specific client machine as input and delivers login tokens as output (and possibly also downloads the CA cert for TLS server validation)

- some logic in client apps to use the primitives inside the login tokens, to establish a client connection. We can build a proof of concept inside `cockroach sql` but eventually we'll want similar logic in Python, a go library, etc.

Jira issue: CRDB-7690

**Comment by jordanlewis** [comment]
cc @petermattis

**Comment by jordanlewis** [comment]
also cc @rafiss.

Excellent writeup of the issue, thank you Raphael!

**Comment by knz** [comment]
See this implementation for a proof of concept: https://github.com/knz/loginserver

**Comment by knz** [comment]
NB: the proof of concept is defective in that it does not use secure tokens, uses a single shared secret, assumes that the crdb server is running at a fixed address, etc, etc. However the general flow of information is similar to what we're expecting for the end user experience.

**Comment by rafiss** [comment]
This needs someone from Product to decide on how to prioritize it, so tagging @kernfeld-cockroach for awareness.

**Comment by knz** [comment]
This will also need support from the seceng team

**Comment by knz** [comment]
cc @jtsiros for triage and tracking

**Comment by data-matt** [comment]
@knz, wondering if you can close this?

**Comment by knz** [comment]
That's a question for @kpatron-cockroachlabs or @jtsiros i think.

**Comment by kpatron-cockroachlabs** [comment]
My bias is to close this issue for two reasons but I defer to Jon. 
* We have incorporated many of the ideas expressed here into the token login flow in 22.2.0.
* If the issue is meant to be owned by CDX, its positioning here is a little awkward for us, since we don't really track things in GitHub. I wasn't aware of this issue until today.

**Comment by knz** [comment]
Fixed in #85965.
